src/librustc/mir/interpret/value.rs

```diff
@@ -92,125 +86,124 @@ impl<'tcx> ty::TypeFoldable<'tcx> for Value {
+                Ok(bits == 0)
+            },
+            Scalar::Ptr(_) => Ok(false),
         }
     }
 
-    pub fn to_value_with_len(self, len: u64) -> Value {
-        Value::ByValPair(self.primval, PrimVal::from_u128(len as u128))
+    pub fn to_value_with_len<C: HasDataLayout>(self, len: u64, cx: C) -> Value {
+        Value::ScalarPair(self, Scalar::Bits {
+            bits: len as u128,
+            defined: cx.data_layout().pointer_size.bits() as u8,
+        })
     }
 
-    pub fn to_value_with_vtable(self, vtable: MemoryPointer) -> Value {
-        Value::ByValPair(self.primval, PrimVal::Ptr(vtable))
+    pub fn to_value_with_vtable(self, vtable: Pointer) -> Value {
+        Value::ScalarPair(self, Scalar::Ptr(vtable))
     }
 
     pub fn to_value(self) -> Value {
-        Value::ByVal(self.primval)
+        Value::Scalar(self)
     }
 }
 
-impl ::std::convert::From<PrimVal> for Pointer {
-    fn from(primval: PrimVal) -> Self {
-        Pointer { primval }
+impl From<Pointer> for Scalar {
+    fn from(ptr: Pointer) -> Self {
+        Scalar::Ptr(ptr)
     }
 }
 
-impl ::std::convert::From<MemoryPointer> for Pointer {
-    fn from(ptr: MemoryPointer) -> Self {
-        PrimVal::Ptr(ptr).into()
-    }
-}
-
-/// A `PrimVal` represents an immediate, primitive value existing outside of a
+/// A `Scalar` represents an immediate, primitive value existing outside of a
 /// `memory::Allocation`. It is in many ways like a small chunk of a `Allocation`, up to 8 bytes in
-/// size. Like a range of bytes in an `Allocation`, a `PrimVal` can either represent the raw bytes
-/// of a simple value, a pointer into another `Allocation`, or be undefined.
+/// size. Like a range of bytes in an `Allocation`, a `Scalar` can either represent the raw bytes
+/// of a simple value or a pointer into another `Allocation`
 #[derive(Clone, Copy, Debug, Eq, PartialEq, RustcEncodable, RustcDecodable, Hash)]
-pub enum PrimVal {
+pub enum Scalar {
     /// The raw bytes of a simple value.
-    Bytes(u128),
+    Bits {
+        /// number of bits that are valid and may be read
+        defined: u8,
+        bits: u128,
+    },
 
     /// A pointer into an `Allocation`. An `Allocation` in the `memory` module has a list of
-    /// relocations, but a `PrimVal` is only large enough to contain one, so we just represent the
-    /// relocation and its associated offset together as a `MemoryPointer` here.
-    Ptr(MemoryPointer),
-
-    /// An undefined `PrimVal`, for representing values that aren't safe to examine, but are safe
-    /// to copy around, just like undefined bytes in an `Allocation`.
-    Undef,
+    /// relocations, but a `Scalar` is only large enough to contain one, so we just represent the
+    /// relocation and its associated offset together as a `Pointer` here.
+    Ptr(Pointer),
 }
 
 #[derive(Clone, Copy, Debug, PartialEq)]
-pub enum PrimValKind {
+pub enum ScalarKind {
     I8, I16, I32, I64, I128,
     U8, U16, U32, U64, U128,
```

Comment:
You could at least wrap `layout::Integer` for signed and unsigned integers, if you want to keep `ScalarKind` at all (as opposed to reusing `layout::Primitive`).

Reply:
done